diff --git a/CRM/Report/Page/InstanceList.php b/CRM/Report/Page/InstanceList.php
index 97657f08297..ae134b7411a 100644
--- a/CRM/Report/Page/InstanceList.php
+++ b/CRM/Report/Page/InstanceList.php
@@ -120,7 +120,7 @@ public function info() {
 
           WHERE v.is_active = 1 {$report}
                 AND inst.domain_id = %9
-          ORDER BY  v.weight ASC, inst.title ASC";
+          ORDER BY  inst.title ASC";
     $queryParams[9] = [CRM_Core_Config::domainID(), 'Integer'];
 
     $dao = CRM_Core_DAO::executeQuery($sql, $queryParams);
