index 4065bed5db3..6508e5e91db 100644
--- a/CRM/Event/Form/Registration/Register.php
+++ b/CRM/Event/Form/Registration/Register.php
@@ -1186,7 +1186,8 @@ public static function checkRegistration($fields, $form, $isAdditional = FALSE)
       }
       $participant->is_test = 0;
       $participant->find();
-      $statusTypes = CRM_Event_PseudoConstant::participantStatus(NULL, 'is_counted = 1');
+      // Event#30 - Anyone whose status type has `is_counted` OR is on the waitlist should be considered as registered.
+      $statusTypes = CRM_Event_PseudoConstant::participantStatus(NULL, 'is_counted = 1') + CRM_Event_PseudoConstant::participantStatus(NULL, "name = 'On waitlist'");
       while ($participant->fetch()) {
         if (array_key_exists($participant->status_id, $statusTypes)) {
           if (!$isAdditional && !$form->_values['event']['allow_same_participant_emails']) {
