diff --git a/src/PhpWord/Shared/Html.php b/src/PhpWord/Shared/Html.php
index c484e08f2b..0c897bb8d8 100644
--- a/src/PhpWord/Shared/Html.php
+++ b/src/PhpWord/Shared/Html.php
@@ -686,7 +686,6 @@ private static function parseImage($node, $element)
                 fclose($ifp);
             }
         }
-        $src = urldecode($src);
 
         if (!is_file($src)
             && !is_null(self::$options)
