diff --git a/CRM/Core/Config.php b/CRM/Core/Config.php
index 86353a8a4..b683955b4 100644
--- a/CRM/Core/Config.php
+++ b/CRM/Core/Config.php
@@ -274,7 +274,9 @@ class CRM_Core_Config extends CRM_Core_Config_MagicMerge {
     UserJob::delete(FALSE)->addWhere('expires_date', '<', 'now')->execute();
     // clear all caches
     self::clearDBCache();
-    Civi::cache('session')->clear();
+    // [ML] SYMBIOTIC This causes problems if people are filling-in quickform
+    // forms during a system.flush
+    // Civi::cache('session')->clear();
     Civi::cache('metadata')->clear();
     CRM_Core_DAO_AllCoreTables::flush();
     CRM_Utils_System::flushCache();
@@ -336,7 +338,9 @@ class CRM_Core_Config extends CRM_Core_Config_MagicMerge {
     $queries = [
       'TRUNCATE TABLE civicrm_acl_cache',
       'TRUNCATE TABLE civicrm_acl_contact_cache',
-      'TRUNCATE TABLE civicrm_cache',
+      // [ML] SYMBIOTIC Reduce risks of losing a quickform session
+      'DELETE FROM civicrm_cache WHERE group_name NOT LIKE "CiviCRM%Session"',
+      // 'TRUNCATE TABLE civicrm_cache',
       'TRUNCATE TABLE civicrm_prevnext_cache',
       'UPDATE civicrm_group SET cache_date = NULL',
       'TRUNCATE TABLE civicrm_group_contact_cache',
