diff --git a/CRM/Utils/Check/Component/Event.php b/CRM/Utils/Check/Component/Event.php
index 1ac9f94d26d..0811ff8c129 100644
--- a/CRM/Utils/Check/Component/Event.php
+++ b/CRM/Utils/Check/Component/Event.php
@@ -5,6 +5,13 @@
  */
 class CRM_Utils_Check_Component_Event extends CRM_Utils_Check_Component {
 
+  /**
+   * @inheritDoc
+   */
+  public function isEnabled() {
+    CRM_Core_Component::isEnabled('CiviEvent');
+  }
+
   /**
    * Check events have timezone set.
    *
