diff --git a/ext/flexmailer/src/Listener/BounceTracker.php b/ext/flexmailer/src/Listener/BounceTracker.php
index 210fea22dfa..e3c3dd11251 100644
--- a/ext/flexmailer/src/Listener/BounceTracker.php
+++ b/ext/flexmailer/src/Listener/BounceTracker.php
@@ -25,6 +25,7 @@ public function onCompose(ComposeBatchEvent $e) {
     }
 
     $mailing = $e->getMailing();
+    $defaultReturnPath = \CRM_Core_BAO_MailSettings::defaultReturnPath();
 
     foreach ($e->getTasks() as $task) {
       /** @var \Civi\FlexMailer\FlexMailerTask $task */
@@ -33,7 +34,7 @@ public function onCompose(ComposeBatchEvent $e) {
         $task->getAddress());
 
       if (!$task->getMailParam('Return-Path')) {
-        $task->setMailParam('Return-Path', $verp['bounce']);
+        $task->setMailParam('Return-Path', $defaultReturnPath ?? $verp['bounce']);
       }
       if (!$task->getMailParam('X-CiviMail-Bounce')) {
         $task->setMailParam('X-CiviMail-Bounce', $verp['bounce']);
