diff --git a/ang/crmUi.js b/ang/crmUi.js
index 6b9a54c622b..309d512bdd6 100644
--- a/ang/crmUi.js
+++ b/ang/crmUi.js
@@ -753,7 +753,7 @@
           if (ctrl.ngModel) {
             // Ensure widget is updated when model changes
             ctrl.ngModel.$render = function() {
-              element.val(ctrl.ngModel.$viewValue || '').change();
+              element.val(ctrl.ngModel.$viewValue || '');
             };
 
             // Copied from ng-list
