commit 1254db53d822f79547af15d7b825b587f86cc09d
Author: nwu on nwu.local <none@megaphonetech.com>
Date:   Fri Aug 4 17:05:56 2023 -0400

    improved #25390 fix

diff --git a/wp-content/plugins/civicrm/civicrm/CRM/Core/Payment/PayPalProIPN.php b/wp-content/plugins/civicrm/civicrm/CRM/Core/Payment/PayPalProIPN.php
index 5b039d3b4..c861c3984 100644
--- a/wp-content/plugins/civicrm/civicrm/CRM/Core/Payment/PayPalProIPN.php
+++ b/wp-content/plugins/civicrm/civicrm/CRM/Core/Payment/PayPalProIPN.php
@@ -260,6 +260,8 @@ class CRM_Core_Payment_PayPalProIPN extends CRM_Core_Payment_BaseIPN {
     recurring_payment_suspended_due_to_max_failed_payment  RP Profile Suspended due to Max Failed Payment
      */
 
+    $recur->processor_id = $this->retrieve('recurring_payment_id', 'String');
+    $recur->trxn_id = $recur->processor_id;
     //set transaction type
     $txnType = $this->retrieve('txn_type', 'String');
     //Changes for paypal pro recurring payment
@@ -268,15 +270,12 @@ class CRM_Core_Payment_PayPalProIPN extends CRM_Core_Payment_BaseIPN {
         if (in_array(CRM_Core_PseudoConstant::getName('CRM_Contribute_BAO_ContributionRecur', 'contribution_status_id', $recur->contribution_status_id), [
           'Pending', 'In Progress',
         ], TRUE)
-          && !empty($recur->processor_id)
         ) {
           echo 'already handled';
           return;
         }
         $recur->create_date = $now;
         $recur->contribution_status_id = CRM_Core_PseudoConstant::getKey('CRM_Contribute_BAO_ContributionRecur', 'contribution_status_id', 'Pending');
-        $recur->processor_id = $this->retrieve('recurring_payment_id', 'String');
-        $recur->trxn_id = $recur->processor_id;
         $subscriptionPaymentStatus = CRM_Core_Payment::RECURRING_PAYMENT_START;
         $sendNotification = TRUE;
         break;
