diff --git a/CRM/Core/Payment/PayPalProIPN.php b/CRM/Core/Payment/PayPalProIPN.php
index 5b039d3b413..27dc131eb95 100644
--- a/CRM/Core/Payment/PayPalProIPN.php
+++ b/CRM/Core/Payment/PayPalProIPN.php
@@ -282,6 +282,8 @@ public function recur(array $input): void {
         break;
 
       case 'recurring_payment':
+        $recur->processor_id = $this->retrieve('recurring_payment_id', 'String');
+        $recur->trxn_id = $recur->processor_id;
         if (!$first) {
           if ($input['paymentStatus'] !== 'Completed') {
             throw new CRM_Core_Exception('Ignore all IPN payments that are not completed');
