diff --git a/Civi/Api4/Action/Contact/GetDuplicates.php b/Civi/Api4/Action/Contact/GetDuplicates.php
index f39298c0034..e3a825d0e2d 100644
--- a/Civi/Api4/Action/Contact/GetDuplicates.php
+++ b/Civi/Api4/Action/Contact/GetDuplicates.php
@@ -147,6 +147,7 @@ public static function fields(BasicGetFieldsAction $action) {
         'action' => 'create',
         'loadOptions' => $action->getLoadOptions(),
         'where' => [['name', 'NOT IN', $ignore], ['type', 'IN', ['Field', 'Custom']]],
+        'checkPermissions' => FALSE,
       ]);
       if ($entity !== 'Contact') {
         $prefix = strtolower($entity) . '_primary.';
