diff --git a/Civi/Api4/Action/MailSettings/TestConnection.php b/Civi/Api4/Action/MailSettings/TestConnection.php
index 51076236b6d..344ef57a3dc 100644
--- a/Civi/Api4/Action/MailSettings/TestConnection.php
+++ b/Civi/Api4/Action/MailSettings/TestConnection.php
@@ -12,6 +12,7 @@
 namespace Civi\Api4\Action\MailSettings;
 
 use Civi\Api4\Generic\BasicBatchAction;
+use Civi\Api4\MailSettings;
 
 class TestConnection extends BasicBatchAction {
 
@@ -24,8 +25,13 @@ public function __construct($entityName, $actionName) {
    * @return array
    */
   protected function doTask($item) {
+    $mailingName = MailSettings::get(FALSE)
+      ->addSelect('name')
+      ->addWhere('id', '=', $item['id'])
+      ->execute()
+      ->first()['name'];
     try {
-      $mailStore = \CRM_Mailing_MailStore::getStore($item['name']);
+      $mailStore = \CRM_Mailing_MailStore::getStore($mailingName);
     }
     catch (\Throwable $t) {
       \Civi::log()->warning('MailSettings: Failed to establish test connection', [
