diff --git a/ext/civigrant/CRM/Grant/BAO/Grant.php b/ext/civigrant/CRM/Grant/BAO/Grant.php
index 0c9916e2db7a..27538b042806 100644
--- a/ext/civigrant/CRM/Grant/BAO/Grant.php
+++ b/ext/civigrant/CRM/Grant/BAO/Grant.php
@@ -110,6 +110,7 @@ public static function add($params, $ids = []) {
       $title = CRM_Contact_BAO_Contact::displayName($grant->contact_id) . ' - ' . ts('Grant') . ': ' . $grantTypes[$grant->grant_type_id];
 
       civicrm_api4('RecentItem', 'create', [
+        'checkPermissions' => FALSE,
         'values' => [
           'entity_type' => 'Grant',
           'entity_id' => $grant->id,
