diff --git a/ext/search_kit/ang/crmSearchDisplay/crmSearchDisplayEditable.component.js b/ext/search_kit/ang/crmSearchDisplay/crmSearchDisplayEditable.component.js
index d54bd66c12e..46eed4615d2 100644
--- a/ext/search_kit/ang/crmSearchDisplay/crmSearchDisplayEditable.component.js
+++ b/ext/search_kit/ang/crmSearchDisplay/crmSearchDisplayEditable.component.js
@@ -11,7 +11,7 @@
       cancel: '&'
     },
     templateUrl: '~/crmSearchDisplay/crmSearchDisplayEditable.html',
-    controller: function($scope, $element, crmApi4) {
+    controller: function($scope, $element, crmApi4, crmStatus) {
       var ctrl = this,
         initialValue,
         col;
@@ -53,7 +53,7 @@
         const value = formatDataType(ctrl.value);
         if (value !== initialValue) {
           col.edit.record[col.edit.value_key] = value;
-          CRM.status({}, crmApi4(col.edit.entity, col.edit.action, {values: col.edit.record}));
+          crmStatus({}, crmApi4(col.edit.entity, col.edit.action, {values: col.edit.record}));
           ctrl.row.data[col.edit.value_path] = value;
           col.val = formatDisplayValue(value);
         }
diff --git a/js/Common.js b/js/Common.js
index 928972794bc..5521a863e71 100644
--- a/js/Common.js
+++ b/js/Common.js
@@ -1320,11 +1320,12 @@ if (!CRM.vars) CRM.vars = {};
       }
     }
     return (deferred || new $.Deferred())
-      .then(function(data) {
+      .done(function(data) {
         // If the server returns an error msg call the error handler
         var status = $.isPlainObject(data) && (data.is_error || data.status === 'error') ? 'error' : 'success';
         handle(status, data);
-      }, function(data) {
+      })
+      .fail(function(data) {
         handle('error', data);
       });
   };
