diff --git a/CRM/Utils/Hook.php b/CRM/Utils/Hook.php
index e96b976ce4f..00e82f3602f 100644
--- a/CRM/Utils/Hook.php
+++ b/CRM/Utils/Hook.php
@@ -1487,7 +1487,7 @@ public static function config(&$config, ?array $flags = NULL) {
       $count++;
     }
     $defaultFlags = ['civicrm' => FALSE, 'uf' => FALSE, 'instances' => $count];
-    $flags = array_merge($defaultFlags, $flags);
+    $flags = !empty($flags) ? array_merge($defaultFlags, $flags) : $defaultFlags;
     $null = NULL;
     return self::singleton()->invoke(['config', 'flags'], $config,
       $flags, $null, $null, $null, $null,
