diff --git a/CRM/Utils/Token.php b/CRM/Utils/Token.php
index 7cd3bf0c38e..c8f1b098541 100644
--- a/CRM/Utils/Token.php
+++ b/CRM/Utils/Token.php
@@ -1798,6 +1798,9 @@ public static function getTokenDeprecations(): array {
           '$participant_status_id' => 'participant.status_id',
 
         ],
+        'event_online_receipt' => [
+          '`$participant.id`' => 'participant.id',
+         ],
         'pledge_acknowledgement' => [
           '$domain' => ts('no longer available / relevant'),
           '$contact' => ts('no longer available / relevant'),
