diff --git a/CRM/Contribute/PseudoConstant.php b/CRM/Contribute/PseudoConstant.php
index 210cc1faa5db..12197d0d8b61 100644
--- a/CRM/Contribute/PseudoConstant.php
+++ b/CRM/Contribute/PseudoConstant.php
@@ -307,34 +307,18 @@ public static function &contributionStatus($id = NULL, $columnName = 'label') {
   }
 
   /**
-   * Get all the Personal campaign pages.
-   *
-   *
-   * @param string|null $pageType
-   * @param int $id
-   *
-   * @return array
-   *   array reference of all pcp if any
+   * Get all the active Personal Campaign Pages.
    */
-  public static function &pcPage($pageType = NULL, $id = NULL) {
-    if (!isset(self::$pcPage[$pageType])) {
-      if ($pageType) {
-        $params = "page_type='{$pageType}'";
-      }
-      else {
-        $params = '';
-      }
-      CRM_Core_PseudoConstant::populate(self::$pcPage[$pageType],
-        'CRM_PCP_DAO_PCP',
-        FALSE, 'title', 'is_active', $params
-      );
-    }
-    $result = self::$pcPage[$pageType];
-    if ($id) {
-      return $result = $result[$id] ?? NULL;
+  public static function &pcPage(): array {
+    if (!isset(self::$pcPage)) {
+      $result = (array) \Civi\Api4\PCP::get(TRUE)
+        ->addSelect('id', 'title')
+        ->addWhere('is_active', '=', TRUE)
+        ->execute()
+        ->indexBy('id');
+      self::$pcPage = array_column($result, 'title', 'id');
     }
-
-    return $result;
+    return self::$pcPage;
   }
 
   /**
