diff --git a/CRM/Event/ICalendar.php b/CRM/Event/ICalendar.php
index 0d16cfe2ad2..7947b7aaf89 100644
--- a/CRM/Event/ICalendar.php
+++ b/CRM/Event/ICalendar.php
@@ -54,10 +54,7 @@ public static function run() {
     $config = CRM_Core_Config::singleton();
 
     $template->assign('events', $info);
-
-    $timezones = [@date_default_timezone_get()];
-
-    $template->assign('timezone', $timezones[0]);
+    $template->assign('timezone', date_default_timezone_get());
 
     // Send data to the correct template for formatting (iCal vs. gData)
     if ($rss) {
diff --git a/CRM/Utils/ICalendar.php b/CRM/Utils/ICalendar.php
index fddefbfec84..01db70f0a52 100644
--- a/CRM/Utils/ICalendar.php
+++ b/CRM/Utils/ICalendar.php
@@ -160,50 +160,42 @@ public static function send($calendar, $content_type = 'text/calendar', $charset
   }
 
   /**
-   * @param array $timezones - Timezone strings
    * @param $date_min
    * @param $date_max
    *
    * @return array
    */
-  public static function generate_timezones(array $timezones, $date_min, $date_max) {
-    if (empty($timezones)) {
-      return [];
-    }
+  public static function generate_timezones($date_min, $date_max) {
+    $tzstr = date_default_timezone_get();
+    $timezone = new DateTimeZone($tzstr);
 
-    $tz_items = [];
+    $transitions = $timezone->getTransitions($date_min, $date_max);
 
-    foreach ($timezones as $tzstr) {
-      $timezone = new DateTimeZone($tzstr);
+    if (count($transitions) === 1) {
+      $transitions[] = array_values($transitions)[0];
+    }
 
-      $transitions = $timezone->getTransitions($date_min, $date_max);
+    $item = [
+      'id' => $timezone->getName(),
+      'transitions' => [],
+    ];
 
-      if (count($transitions) === 1) {
-        $transitions[] = array_values($transitions)[0];
-      }
+    $last_transition = array_shift($transitions);
 
-      $item = [
-        'id' => $timezone->getName(),
-        'transitions' => [],
+    foreach ($transitions as $transition) {
+      $item['transitions'][] = [
+        'type' => $transition['isdst'] ? 'DAYLIGHT' : 'STANDARD',
+        'offset_from' => self::format_tz_offset($last_transition['offset']),
+        'offset_to' => self::format_tz_offset($transition['offset']),
+        'abbr' => $transition['abbr'],
+        'dtstart' => date_create($transition['time'], $timezone)->format("Ymd\THis"),
       ];
 
-      $last_transition = array_shift($transitions);
-
-      foreach ($transitions as $transition) {
-        $item['transitions'][] = [
-          'type' => $transition['isdst'] ? 'DAYLIGHT' : 'STANDARD',
-          'offset_from' => self::format_tz_offset($last_transition['offset']),
-          'offset_to' => self::format_tz_offset($transition['offset']),
-          'abbr' => $transition['abbr'],
-          'dtstart' => date_create($transition['time'], $timezone)->format("Ymd\THis"),
-        ];
-
-        $last_transition = $transition;
-      }
-
-      $tz_items[] = $item;
+      $last_transition = $transition;
     }
 
+    $tz_items[] = $item;
+
     return $tz_items;
   }
 
@@ -245,7 +237,7 @@ public static function createCalendarFile($info) {
           return strtotime($event['end_date'] ?? $event['start_date']);
         }, $info)
       );
-      $template->assign('timezones', CRM_Utils_ICalendar::generate_timezones($timezones, $date_min, $date_max));
+      $template->assign('timezones', CRM_Utils_ICalendar::generate_timezones($date_min, $date_max));
     }
     else {
       $template->assign('timezones', NULL);
