diff --git a/CRM/Contact/BAO/Query.php b/CRM/Contact/BAO/Query.php
index 69cdae13458..1b169ccc1e4 100644
--- a/CRM/Contact/BAO/Query.php
+++ b/CRM/Contact/BAO/Query.php
@@ -4495,6 +4495,7 @@ public static function &defaultReturnProperties($mode = 1) {
           'legal_name' => 1,
           'sic_code' => 1,
           'current_employer' => 1,
+          'contact_source' => 1,
           // FIXME: should we use defaultHierReturnProperties() for the below?
           'do_not_email' => 1,
           'do_not_mail' => 1,
