diff --git a/CRM/Member/BAO/Membership.php b/CRM/Member/BAO/Membership.php
index 79015e14a1d..e8d838957ec 100644
--- a/CRM/Member/BAO/Membership.php
+++ b/CRM/Member/BAO/Membership.php
@@ -163,11 +163,13 @@ public static function add(&$params) {
       }
 
       foreach (['Membership Signup', 'Membership Renewal'] as $activityType) {
-        $activityParams['id'] = civicrm_api3('Activity', 'Get', [
-          'source_record_id' => $membership->id,
-          'activity_type_id' => $activityType,
-          'status_id' => 'Scheduled',
-        ])['id'] ?? NULL;
+        $activityParams['id'] = \Civi\Api4\Activity::get(FALSE)
+          ->addSelect('id')
+          ->addWhere('source_record_id', '=', $membership->id)
+          ->addWhere('activity_type_id:name', '=', $activityType)
+          ->addWhere('status_id:name', '=', 'Scheduled')
+          ->execute()
+          ->first()['id'] ?? NULL;
         // 1. Update Schedule Membership Signup/Renwal activity to completed on successful payment of pending membership
         // 2. OR Create renewal activity scheduled if its membership renewal will be paid later
         if (!empty($params['membership_activity_status']) && (!empty($activityParams['id']) || $activityType == 'Membership Renewal')) {
