diff --git a/CRM/Utils/REST.php b/CRM/Utils/REST.php
index 9e334bab42e1..ed738c0b210d 100644
--- a/CRM/Utils/REST.php
+++ b/CRM/Utils/REST.php
@@ -323,6 +323,10 @@ public static function &buildParamList() {
         ]);
       }
     }
+    // Handle POST requests of content-type application/json.
+    elseif (array_key_exists('json', $requestParams) && is_array($requestParams['json'])) {
+      $params = $requestParams['json'];
+    }
     foreach ($requestParams as $n => $v) {
       if (!array_key_exists($n, $skipVars)) {
         $params[$n] = $v;
diff --git a/CRM/Utils/Request.php b/CRM/Utils/Request.php
index 0c60dd710d7e..a0994488ddc7 100644
--- a/CRM/Utils/Request.php
+++ b/CRM/Utils/Request.php
@@ -133,6 +133,16 @@ protected static function getValue($name, $method) {
     foreach (($method ?? []) as $key => $value) {
       if (strpos($key, 'amp;') !== FALSE) {
         $method[str_replace('amp;', '', $key)] = $method[$key];
+        if ($method === '$_POST' && $_SERVER['CONTENT_TYPE'] === 'application/json') {
+          static $post = NULL;
+          if (!isset($post)) {
+            $rawPost = file_get_contents('php://input');
+            $post = json_decode($rawPost, TRUE) ?? [];
+          }
+          if (isset($post[$name])) {
+            return $post[$name];
+          }
+        }
         if (isset($method[$name])) {
           return $method[$name];
         }
@@ -167,6 +177,11 @@ public static function exportValues() {
     if ($_POST) {
       $result = array_merge($result, $_POST);
     }
+    if ($_SERVER['REQUEST_METHOD'] === 'POST' && $_SERVER['CONTENT_TYPE'] === 'application/json') {
+      $rawPost = file_get_contents('php://input');
+      $jsonPost = json_decode($rawPost, TRUE) ?? [];
+      $result = array_merge($result, $jsonPost);
+    }
     return $result;
   }
 
