diff --git a/ext/civigrant/CRM/Grant/BAO/Query.php b/ext/civigrant/CRM/Grant/BAO/Query.php
index 977d64d5b41a..d1c3227afc25 100644
--- a/ext/civigrant/CRM/Grant/BAO/Query.php
+++ b/ext/civigrant/CRM/Grant/BAO/Query.php
@@ -304,6 +304,7 @@ public static function getSearchFieldMetadata() {
       'grant_decision_date',
       'grant_money_transfer_date',
       'grant_due_date',
+      'grant_status_id',
     ];
     $metadata = civicrm_api3('Grant', 'getfields', [])['values'];
     return array_intersect_key($metadata, array_flip($fields));
diff --git a/ext/civigrant/CRM/Grant/Form/Search.php b/ext/civigrant/CRM/Grant/Form/Search.php
index c59c1875d9f8..ee1a156d0526 100644
--- a/ext/civigrant/CRM/Grant/Form/Search.php
+++ b/ext/civigrant/CRM/Grant/Form/Search.php
@@ -221,12 +221,6 @@ public function fixFormValues() {
       return;
     }
 
-    $status = CRM_Utils_Request::retrieve('status', 'String');
-    if ($status) {
-      $this->_formValues['grant_status_id'] = $status;
-      $this->_defaults['grant_status_id'] = $status;
-    }
-
     $cid = CRM_Utils_Request::retrieve('cid', 'Positive', $this);
 
     if ($cid) {
diff --git a/ext/civigrant/templates/CRM/Grant/Page/DashBoard.tpl b/ext/civigrant/templates/CRM/Grant/Page/DashBoard.tpl
index bf3fadba0151..dbbbe57c64d8 100644
--- a/ext/civigrant/templates/CRM/Grant/Page/DashBoard.tpl
+++ b/ext/civigrant/templates/CRM/Grant/Page/DashBoard.tpl
@@ -30,8 +30,8 @@ You have {$grantSummary.total_grants} grant(s) registered in your database.
 
 {foreach from=$grantSummary.per_status item=status key=id}
 <tr>
-    <td><a href="{crmURL p="civicrm/grant/search" q="reset=1&status=`$id`&force=1"}">{$status.label}</a></td>
-    <td><a href="{crmURL p="civicrm/grant/search" q="reset=1&status=`$id`&force=1"}">{$status.total}</a></td>
+    <td><a href="{crmURL p="civicrm/grant/search" q="reset=1&grant_status_id=`$id`&force=1"}">{$status.label}</a></td>
+    <td><a href="{crmURL p="civicrm/grant/search" q="reset=1&grant_status_id=`$id`&force=1"}">{$status.total}</a></td>
 </tr>
 {/foreach}
 <tr class="columnfooter">
