diff --git a/Civi/WorkflowMessage/Traits/CustomFieldTrait.php b/Civi/WorkflowMessage/Traits/CustomFieldTrait.php
index 06a5166c3782..5dc81eeecb82 100644
--- a/Civi/WorkflowMessage/Traits/CustomFieldTrait.php
+++ b/Civi/WorkflowMessage/Traits/CustomFieldTrait.php
@@ -91,9 +91,9 @@ protected function getCustomFieldDisplay(array $entityRecord, string $entity, ar
     foreach ($viewableFields as $fieldSpec) {
       $fieldName = $fieldSpec['custom_group_id.name'] . '.' . $fieldSpec['name'];
       $value = str_replace('&nbsp;', '', \CRM_Core_BAO_CustomField::displayValue($entityRecord[$fieldName], $fieldSpec['id'], $entityRecord['id']));
-      // I can't see evidence we have filtered out empty strings here historically
-      // but maybe we should?
-      $fields[$fieldSpec['custom_group_id.title']][$fieldSpec['label']] = $value;
+      if (!empty($value)) {
+        $fields[$fieldSpec['custom_group_id.title']][$fieldSpec['label']] = $value;
+      }
     }
     return $fields;
   }
