diff --git a/ext/search_kit/Civi/Api4/Action/SearchDisplay/Run.php b/ext/search_kit/Civi/Api4/Action/SearchDisplay/Run.php
index 160ad0eac8bb..343f4f8598dd 100644
--- a/ext/search_kit/Civi/Api4/Action/SearchDisplay/Run.php
+++ b/ext/search_kit/Civi/Api4/Action/SearchDisplay/Run.php
@@ -181,6 +181,9 @@ private function getTally(): array {
         $select[] = $sqlFnClass::renderExpression(implode(' ', $fnArgs)) . " `$tallyKey`";
       }
     }
+    if (!$select) {
+      return [];
+    }
     $query = 'SELECT ' . implode(', ', $select) . "\nFROM (" . $sql . ")\n`api_query`";
     $dao = \CRM_Core_DAO::executeQuery($query);
     $dao->fetch();
diff --git a/ext/search_kit/ang/crmSearchAdmin/displays/searchAdminDisplayTable.component.js b/ext/search_kit/ang/crmSearchAdmin/displays/searchAdminDisplayTable.component.js
index 9301378515f9..0341ba8a1373 100644
--- a/ext/search_kit/ang/crmSearchAdmin/displays/searchAdminDisplayTable.component.js
+++ b/ext/search_kit/ang/crmSearchAdmin/displays/searchAdminDisplayTable.component.js
@@ -59,6 +59,9 @@
         // and populate or validate this.settings.columns
         if (value !== 'auto') {
           this.parent.initColumns({label: true, sortable: true});
+          if (this.display.settings.tally) {
+            this.setTallyDefaults();
+          }
         }
         this.display.settings.columnMode = value;
       };
@@ -93,22 +96,26 @@
         }
       };
 
-      this.toggleTally = function() {
-        if (ctrl.display.settings.tally) {
-          delete ctrl.display.settings.tally;
-          ctrl.display.settings.columns.forEach((col) => delete col.tally);
+      this.toggleTally = () => {
+        if (this.display.settings.tally) {
+          delete this.display.settings.tally;
+          this.display.settings.columns.forEach((col) => delete col.tally);
         } else {
-          ctrl.display.settings.tally = {label: ts('Total')};
-          ctrl.display.settings.columns.forEach(function(col) {
-            if (col.type === 'field') {
-              col.tally = {
-                fn: searchMeta.getDefaultAggregateFn(searchMeta.parseExpr(ctrl.parent.getExprFromSelect(col.key)), ctrl.apiParams)
-              };
-            }
-          });
+          this.display.settings.tally = {label: ts('Total')};
+          this.setTallyDefaults();
         }
       };
 
+      this.setTallyDefaults = () => {
+        this.display.settings.columns?.forEach((col) => {
+          if (col.type === 'field') {
+            col.tally = {
+              fn: searchMeta.getDefaultAggregateFn(searchMeta.parseExpr(this.parent.getExprFromSelect(col.key)), this.apiParams)
+            };
+          }
+        });
+      };
+
       this.getTallyFunctions = function() {
         const allowedFunctions = CRM.crmSearchAdmin.functions.filter((fn) => fn.category === 'aggregate' && fn.params.length);
         return {results: formatForSelect2(allowedFunctions, 'name', 'title', ['description'])};
