diff --git a/ext/search_kit/ang/crmSearchTasks/crmSearchTasksMenu.component.js b/ext/search_kit/ang/crmSearchTasks/crmSearchTasksMenu.component.js
index ae60b838044d..d33b114bf668 100644
--- a/ext/search_kit/ang/crmSearchTasks/crmSearchTasksMenu.component.js
+++ b/ext/search_kit/ang/crmSearchTasks/crmSearchTasksMenu.component.js
@@ -13,6 +13,9 @@
         ctrl = this;
 
       this.$onInit = function() {
+        if (ctrl.displayMode === 'buttons') {
+          ctrl.taskManager.getMetadata();
+        }
         // When a row is selected for bulk actions, load the actions menu
         let unwatchIDs = $scope.$watch('$ctrl.ids.length', function (idsLength) {
           if (idsLength) {
