commit f851ba75413cd51da887d0d06d3016c747e43cca
Author: cprj on cprj.local <none@megaphonetech.com>
Date:   Thu Mar 20 17:38:50 2025 -0400

    improved fix for civirules

diff --git a/civirules/CRM/Civirules/Action.php b/civirules/CRM/Civirules/Action.php
index 8e4d74cb7..bc3f15f7c 100644
--- a/civirules/CRM/Civirules/Action.php
+++ b/civirules/CRM/Civirules/Action.php
@@ -109,7 +109,7 @@ abstract class CRM_Civirules_Action {
    */
   abstract public function getExtraDataInputUrl($ruleActionId);
 
-  public function getFormattedExtraDataInputUrl($url, $ruleActionID) {
+  public function getFormattedExtraDataInputUrl(string $url, int $ruleActionID) {
     return CRM_Utils_System::url($url, 'rule_action_id=' . $ruleActionID, FALSE, NULL, FALSE, FALSE, TRUE);
   }
 
