diff --git a/CRM/Contribute/Form/Contribution/Confirm.php b/CRM/Contribute/Form/Contribution/Confirm.php
index e506bc972ed..10ec56b92eb 100644
--- a/CRM/Contribute/Form/Contribution/Confirm.php
+++ b/CRM/Contribute/Form/Contribution/Confirm.php
@@ -2189,6 +2189,7 @@ protected function processFormSubmission($contactID) {
 
     // store the fact that this is a membership and membership type is selected
     if ($this->isMembershipSelected()) {
+      $result = [];
       $this->doMembershipProcessing($contactID, $membershipParams, $premiumParams);
     }
     else {
@@ -2248,8 +2249,11 @@ protected function processFormSubmission($contactID) {
           }
         }
       }
-      return $result;
     }
+    if (empty($result['is_payment_failure'])) {
+      $this->createCMSUser($contactID);
+    }
+    return $result;
   }
 
   /**
@@ -2378,14 +2382,6 @@ protected function doMembershipProcessing($contactID, $membershipParams, $premiu
         CRM_Core_Session::singleton()->setStatus($e->getMessage());
         CRM_Utils_System::redirect(CRM_Utils_System::url('civicrm/contribute/transact', "_qf_Main_display=true&qfKey={$this->_params['qfKey']}"));
       }
-      if (!$this->_amount > 0.0 || !$membershipParams['amount']) {
-        // we need to explicitly create a CMS user in case of free memberships
-        // since it is done under processConfirm for paid memberships
-        CRM_Contribute_BAO_Contribution_Utils::createCMSUser($membershipParams,
-          $membershipParams['cms_contactID'],
-          'email-' . $this->_bltID
-        );
-      }
     }
   }
 
@@ -2554,18 +2550,6 @@ public function processConfirm(
         $form->_bltID,
         $isRecur
       );
-      // CRM-13074 - create the CMSUser after the transaction is completed as it
-      // is not appropriate to delete a valid contribution if a user create problem occurs
-      if (isset($this->_params['related_contact'])) {
-        $contactID = $this->_params['related_contact'];
-      }
-      elseif (isset($this->_params['cms_contactID'])) {
-        $contactID = $this->_params['cms_contactID'];
-      }
-      CRM_Contribute_BAO_Contribution_Utils::createCMSUser($this->_params,
-        $contactID,
-        'email-' . $form->_bltID
-      );
 
       $paymentParams['item_name'] = $form->_params['description'];
 
@@ -2646,6 +2630,21 @@ public function processConfirm(
     throw new CRM_Core_Exception('code is unreachable, exception is for clarity for refactoring');
   }
 
+  private function createCMSUser($contactID) : void {
+    // CRM-13074 - create the CMSUser after the transaction is completed as it
+    // is not appropriate to delete a valid contribution if a user create problem occurs
+    if (isset($this->_params['related_contact'])) {
+      $contactID = $this->_params['related_contact'];
+    }
+    elseif (isset($this->_params['cms_contactID'])) {
+      $contactID = $this->_params['cms_contactID'];
+    }
+    CRM_Contribute_BAO_Contribution_Utils::createCMSUser($this->_params,
+      $contactID,
+      'email-' . $this->_bltID
+    );
+  }
+
   /**
    * Temporary function to allow unit tests to access function being refactored away.
    *
