diff --git a/CRM/Contact/Form/Edit/TagsAndGroups.php b/CRM/Contact/Form/Edit/TagsAndGroups.php
index 73e9420dff..5fcbddb2ce 100644
--- a/CRM/Contact/Form/Edit/TagsAndGroups.php
+++ b/CRM/Contact/Form/Edit/TagsAndGroups.php
@@ -103,11 +103,11 @@ class CRM_Contact_Form_Edit_TagsAndGroups {
           }
 
           if ($groupElementType == 'select') {
-            $groupsOptions[$id] = $group;
+            $groupsOptions[$group['id']] = $group;
           }
           else {
-            $form->_tagGroup[$fName][$id]['description'] = $group['description'];
-            $elements[] = &$form->addElement('advcheckbox', $id, NULL, $group['text'], $attributes);
+            $form->_tagGroup[$fName][$group['id']]['description'] = $group['description'];
+            $elements[] = &$form->addElement('advcheckbox', $group['id'], NULL, $group['text'], $attributes);
           }
         }
 
