diff --git a/CRM/Core/JobManager.php b/CRM/Core/JobManager.php
index ed2954fc1d1..36d67c21228 100644
--- a/CRM/Core/JobManager.php
+++ b/CRM/Core/JobManager.php
@@ -36,6 +36,11 @@ class CRM_Core_JobManager {
 
   public $_source = NULL;
 
+  /**
+   * @var string The time the cron job started.
+   */
+  private $cronStartTime = NULL;
+
   /**
    * Class constructor.
    */
@@ -43,6 +48,9 @@ public function __construct() {
     $config = CRM_Core_Config::singleton();
     $config->fatalErrorHandler = 'CRM_Core_JobManager_scheduledJobFatalErrorHandler';
 
+    // Save the cron job start time to prevent start time slippage
+    // We force the seconds to 00 because a) cron can't run at second intervals, b) race conditions.
+    $this->cronStartTime = date('YmdHi') . '00';
     $this->jobs = $this->_getJobs();
   }
 
@@ -117,7 +125,7 @@ public function executeJob($job) {
     }
 
     $this->logEntry('Starting execution of ' . $job->name);
-    $job->saveLastRun();
+    $job->saveLastRun($this->cronStartTime);
 
     $singleRunParamsKey = strtolower($job->api_entity . '_' . $job->api_action);
 
diff --git a/CRM/Core/ScheduledJob.php b/CRM/Core/ScheduledJob.php
index fbf0bfd5b84..a6d362cceab 100644
--- a/CRM/Core/ScheduledJob.php
+++ b/CRM/Core/ScheduledJob.php
@@ -61,12 +61,12 @@ public function __construct($params) {
   }
 
   /**
-   * @param null $date
+   * @param string $date a timestamp string in YmdHis format
    */
   public function saveLastRun($date = NULL) {
     $dao = new CRM_Core_DAO_Job();
     $dao->id = $this->id;
-    $dao->last_run = ($date == NULL) ? CRM_Utils_Date::currentDBDate() : CRM_Utils_Date::currentDBDate($date);
+    $dao->last_run = $date ?? CRM_Utils_Date::currentDBDate();
     $dao->save();
   }
 
