diff --git a/vendor/pear/log/php-8.1-strftime.php b/vendor/pear/log/php-8.1-strftime.php
index 0f84da6e9..06c2e6fe7 100644
--- a/vendor/pear/log/php-8.1-strftime.php
+++ b/vendor/pear/log/php-8.1-strftime.php
@@ -102,6 +102,11 @@
       $calendar = IntlGregorianCalendar::createInstance();
       $calendar->setGregorianChange(PHP_INT_MIN);
 
+      // AW 2023-10-29 Temporary fix for compatibility with PHP 8.1.25
+      if ($locale == 'C') {
+          $locale = 'en-US';
+      }
+
       return (new IntlDateFormatter($locale, $date_type, $time_type, $tz, $calendar, $pattern))->format($timestamp);
     };
 
