diff --git a/vendor/pear/mail/Mail/smtp.php b/vendor/pear/mail/Mail/smtp.php
index 5e698feee..f171c088d 100644
--- a/vendor/pear/mail/Mail/smtp.php
+++ b/vendor/pear/mail/Mail/smtp.php
@@ -289,9 +289,11 @@ class Mail_smtp extends Mail {
         /* Since few MTAs are going to allow this header to be forged
          * unless it's in the MAIL FROM: exchange, we'll use
          * Return-Path instead of From: if it's set. */
+/*
         if (!empty($headers['Return-Path'])) {
             $from = $headers['Return-Path'];
         }
+*/
 
         if (!isset($from)) {
             $this->_smtp->rset();
